verification/article/article.md

```diff
@@ -0,0 +1,1063 @@
+   properties, such as monotonicity and round-trip safety.
+3. **`Config`, `Investments`:** The data modeling layer. These modules define the primary data structures of the system,
+   including the main `Config` datatype which encapsulates all sale parameters and rules.
+4. **`Deposit`:** The workflow specification layer. This module composes the primitives from lower layers to define the
+   complete, pure specification of a user deposit, including complex refund logic.
+5. **`Launchpad`:** The top-level state machine. This module defines the complete contract state and models the state
+   transitions (e.g., handling a deposit) by invoking the specifications from the `Deposit` module.
+
+A key pattern employed throughout the codebase is the **Specification-Implementation Separation** [@leino2017tutorial].
+For most critical
+operations, a `function` ending in `...Spec` defines the pure mathematical contract. This allows us to reason about the
+system's logic at an abstract, mathematical
+level.
+
+## 3. Foundational Layer: Verification of Non-Linear Integer Arithmetic
+
+Reasoning about the multiplication and division of integers is a well-known challenge in automated
+verification [@monniaux2008pitfalls; @audemard2021certified]. SMT
+solvers are highly effective for linear integer arithmetic, but non-linear properties often require explicit proof
+guidance. The `MathLemmas` module provides this guidance by establishing a set of trusted, reusable axioms for the rest
+of the system.
+
+The core of financial calculations in this system is scaling a value $x$ by a rational factor $y/k$, implemented using
+integer arithmetic as $(x \cdot y) / k$. The following key lemmas were proven from first principles:
+
+* **Monotonicity with Favorable Scaling (`Lemma_MulDivGreater_From_Scratch`):** This lemma proves that if the scaling
+  factor is greater than or equal to 1, the result is no less than the original amount.
+
+    * **Property 1.** $\forall x, y, k \in \mathbb{N}$ where $x > 0$, $k > 0$, and $y \ge k$
+      $$\frac{x \cdot y}{k} \ge x$$
+      This is crucial for proving that conversions at a stable or favorable price do not result in a loss of principal.
+
+* **Strict Monotonicity with Highly Favorable Scaling (`Lemma_MulDivStrictlyGreater_From_Scratch`):** Due to integer
+  division truncation, $y > k$ is insufficient to guarantee $(x \cdot y) / k > x$. This lemma establishes a stronger
+  precondition to ensure a strict increase.
+
+    * **Property 2.** $\forall x, y, k \in \mathbb{N}$ where $x > 0$, $k > 0$, and $y \ge 2k$:\
+      $$\frac{x \cdot y}{k} > x$$
+      This is used to prove that a significantly favorable price or a large bonus yields a tangible gain.
+
+* **Round-trip Truncation (`Lemma_DivMul_LTE`):** This lemma formalizes the fundamental property of integer division:
+  information may be lost.
+
+    * **Property 3.** $\forall x, y \in \mathbb{N}$ where $y > 0$:
+      $$ \left( \frac{x}{y} \right) \cdot y \le x $$
```

Comment:
Possible future work:

We could prove an even stronger property here. We know by the Euclidean algorithm we can always write `x = ky + r` for some `k` and some `0 <= r < y`. Using these values `k` and `r`, we know the result of `(x/y) * y` is equal to `(k + r/y)*y`. Since `r < y` and integer division always rounds down, the second term is 0 and thus the final result is `(x/y) * y = ky`. Hence we see that `x - (x/y)*y = x - ky = r` (from the original definition of `k` and `r`).

Thus we can prove the stronger property that `0 <= x - (x/y) * y < y` (the current lemma only has the lower bound of this inequality). Including the upper bound is useful because it bound users can experience in round trip asset conversions.

Reply:
Resolved in #76